crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -8022,6 +8022,9 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                                     FileScopeId::global(),
                                     ConstraintKey::NarrowingConstraint(constraint),
                                 ));
+                                // Reaching here means that no bindings are found in any scope.
+                                // Since `explicit_global_symbol` may return a cycle initial value, we return `Place::Undefined` here.
+                                return Place::Undefined.into();
```

Comment:
If returning here is correct, then pushing the constraint into `constraint_keys` immediately above this cannot have any effect, so should we remove that as well? (Removing it doesn't seem to fail any test.)

Why do we hit the `FoundConstraint` case in this situation in the first place? It's not clear to me where the "constraint" comes from.

Reply:
Since place includes member as well as symbol, narrowing constraints must still be tracked even if there are no bindings, for example:

```python
from typing import Literal

class Foo:
    bar: Literal[1, 2]

foo = Foo()
if foo.bar == 1:
    # `foo.bar` has no explicit binding, but the constraint is `foo.bar == 1`
    [reveal_type(foo.bar) for _ in ()]  # Literal[1]
```

But as you pointed out, there was no need to add the constraint to `constraint_keys` here, because we already added it when iterating over `ancestor_scopes` above. Fixed.

Reply:
A more fundamental issue is that the global scope should be excluded from the scopes of enclosing scope lookup.
See the comments below.